### PR TITLE
Add ZAP templates to src/app

### PR DESCRIPTION
**Pull request by jepenven-silabs**
#### Problem
PR #3464 is huge. Break down into multiples small PR to facilitate review.

 #### Summary of Changes
Add Zap Templates in the CHIP repo

 Partial Fix of  #1733

### CODE REVIEW

File: third_party/zap/templates/client-command-macro.zapt
Comment (vivien-apple) [inline third_party/zap/templates/client-command-macro.zapt:3]: Since this file is unused at the moment, maybe we don't even need to include it ?

File: third_party/zap/templates/endpoint_config.zapt
Comment (vivien-apple) [inline third_party/zap/templates/endpoint_config.zapt:4]: Since this file is unused at the moment maybe we don't even need to include it ?

File: third_party/zap/templates/gen_config.zapt
Comment (vivien-apple) [inline third_party/zap/templates/gen_config.zapt:4]: nit: add a line break

File: third_party/zap/templates/helper-chip.js
Comment (vivien-apple) [inline third_party/zap/templates/helper-chip.js:67]: Should this function name be `isEnabled` instead ?

File: third_party/zap/templates/helper-chip.js
Comment (vivien-apple) [inline third_party/zap/templates/helper-chip.js:79]: nit: That's just preference. But since those if blocks just return true if the are called I would use:

```
if (0 == clusterSide.localeCompare(source)) {
    return false;
}

if (isClient(clusterSide) && outgoing) {
    return true;
}

if (isServer(clusterSide) && incoming) {
    return true;
}

return false;
```

But feel free to ignore.

File: third_party/zap/templates/print-cluster.zapt
Comment (vivien-apple) [inline third_party/zap/templates/print-cluster.zapt:15]: Should we replace `SILABS` by `CHIP` in this file ? It's not a big deal since my understanding of this code is that the macro names are just used to generate `CLUSTER_IDS_TO_NAMES` which is used itself by `src/app/util/util.cpp` but I'm wondering since I believe the original idea was to prefix the macro to avoid name collision.

**Review by vivien-apple — COMMENTED** [review]
Looks pretty good. A few minor comments/nits mostly related to style consistency.

My biggest concerns would be that I would really prefer those templates to live inside `examples`. It would make it extra clear that they are used to generate the examples and are basically glued to them.

**Comment by vivien-apple** [comment]
@woody-apple **Restyled** seems to fail because of the Javascript file included in this PR (third_party/repo/templates/helper-chip.js). Do you know what is the proper way to fix it ?

File: third_party/zap/templates/README.md
Comment (jepenven-silabs) [reply third_party/zap/templates/README.md:26]: Done

File: third_party/zap/templates/README.md
Comment (jepenven-silabs) [reply third_party/zap/templates/README.md:22]: Adding a script is a good idea, however I would wait until all ZAP PR land and all templates are completed to do so.

File: third_party/zap/templates/endpoint_config.zapt
Comment (jepenven-silabs) [reply third_party/zap/templates/endpoint_config.zapt:4]: I would keep it as a place holder. Also it tells people that the template is to be completed in a near future.

File: third_party/zap/templates/helper-chip.js
Comment (jepenven-silabs) [reply third_party/zap/templates/helper-chip.js:67]: done

File: third_party/zap/templates/helper-chip.js
Comment (jepenven-silabs) [reply third_party/zap/templates/helper-chip.js:79]: Difference of opinion here, I prefer having checks that are related inside a `if {} else if`

File: third_party/zap/templates/print-cluster.zapt
Comment (jepenven-silabs) [reply third_party/zap/templates/print-cluster.zapt:15]: done

**Comment by jepenven-silabs** [comment]
> Looks pretty good. A few minor comments/nits mostly related to style consistency.
> 
> My biggest concerns would be that I would really prefer those templates to live inside `examples`. It would make it extra clear that they are used to generate the examples and are basically glued to them.

My point of view is that templates are an extension of ZAP. By themselves, they're useless. Moving them elsewhere would give the impression that they can be "standalone".  And yes, as for now only the examples use them, but when people will start using CHIP for their own project, having the templates hidden under `/examples` won't make that much sense anymore.

**Comment by rwalker-apple** [comment]
> > Looks pretty good. A few minor comments/nits mostly related to style consistency.
> > My biggest concerns would be that I would really prefer those templates to live inside `examples`. It would make it extra clear that they are used to generate the examples and are basically glued to them.
> 
> My point of view is that templates are an extension of ZAP. By themselves, they're useless. Moving them elsewhere would give the impression that they can be "standalone". And yes, as for now only the examples use them, but when people will start using CHIP for their own project, having the templates hidden under `/examples` won't make that much sense anymore.

I think these belong under src/app, as they are part of CHIP, they're effectively libraries/source files for CHIP's interaction with ZAP.

**Review by rwalker-apple — CHANGES_REQUESTED** [review]
please move to src/app/zapt/xxx.zapt ?

**Comment by jepenven-silabs** [comment]
> please move to src/app/zapt/xxx.zapt ?
@rwalker-apple 

Only the templates files or everything that is under `third_party/zap/templates/` ?

**Comment by mspang** [comment]
> > > Looks pretty good. A few minor comments/nits mostly related to style consistency.
> > > My biggest concerns would be that I would really prefer those templates to live inside `examples`. It would make it extra clear that they are used to generate the examples and are basically glued to them.
> > 
> > 
> > My point of view is that templates are an extension of ZAP. By themselves, they're useless. Moving them elsewhere would give the impression that they can be "standalone". And yes, as for now only the examples use them, but when people will start using CHIP for their own project, having the templates hidden under `/examples` won't make that much sense anymore.
> 
> I think these belong under src/app, as they are part of CHIP, they're effectively libraries/source files for CHIP's interaction with ZAP.

+1

These files 

> > > Looks pretty good. A few minor comments/nits mostly related to style consistency.
> > > My biggest concerns would be that I would really prefer those templates to live inside `examples`. It would make it extra clear that they are used to generate the examples and are basically glued to them.
> > 
> > 
> > My point of view is that templates are an extension of ZAP. By themselves, they're useless. Moving them elsewhere would give the impression that they can be "standalone". And yes, as for now only the examples use them, but when people will start using CHIP for their own project, having the templates hidden under `/examples` won't make that much sense anymore.
> 
> I think these belong under src/app, as they are part of CHIP, they're effectively libraries/source files for CHIP's interaction with ZAP.

I agree with moving them.

IMO, the things that belongs in third_party/X are 
- additions that could be upstream (like adding a build file for a different build system)
- dependency readme and other administrivia (scripts to update it, etc)

For projects that add tools supporting a domain specific language such as ZAP, the sources in that language don't go in third_party. They are go in the regular directory structure of CHIP.

File: third_party/zap/templates/att-storage.zapt
Comment (jepenven-silabs) [reply third_party/zap/templates/att-storage.zapt]: Done

File: src/app/docs/README.md
Comment (vivien-apple) [inline src/app/docs/README.md:32]: `### <code>/src/app/zap></code>` -> `### <code>/src/app/zap-templates</code>`

If the intent of this folder is to only contains templates, then I would prefer `zap-templates` since `zap` really sounds like the software itself.

And don't forget to fix the extra `>` :)

File: src/app/zap/gen_config.zapt
Comment (vivien-apple) [inline src/app/zap/gen_config.zapt:17]: We are getting rid of the files that uses this macro in #3688 (and this code was not used anyway). So we can probably get rid of this macro here as well.

**Comment by vivien-apple** [comment]
> @woody-apple **Restyled** seems to fail because of the Javascript file included in this PR (third_party/repo/templates/helper-chip.js). Do you know what is the proper way to fix it ?

I opened #3708 for that. The issue being that `clang-format` is trying to parse `helper-chip.js` but `.clang-format` does not contain any rules about JS files...

**Review by vivien-apple — APPROVED** [review]

File: src/app/zap-templates/af-structs.zapt
Comment (bzbarsky-apple) [inline src/app/zap-templates/af-structs.zapt:4]: `#pragma once`?  And similar for other headers?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/zap-templates/af-structs.zapt
Comment (jepenven-silabs) [reply src/app/zap-templates/af-structs.zapt:4]: Done

**Review by rwalker-apple — APPROVED** [review]
didn't want to use the name src/app/zapt?

**Comment by jepenven-silabs** [comment]
> didn't want to use the name src/app/zapt?

If the folder was named zapt one could expect to only find `*.zapt` file inside. Also @vivien-apple requested the zap-templates.

**Comment by vivien-apple** [comment]
> > didn't want to use the name src/app/zapt?
> 
> If the folder was named zapt one could expect to only find `*.zapt` file inside. Also @vivien-apple requested the zap-templates.

This is exactly why I asked for `zap-templates` instead of `zapt`.

**Comment by jepenven-silabs** [comment]
@mspang @andy31415 so we can merge this PR.

**Review by BroderickCarlin — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Pull request merged**
